### PR TITLE
Bug: Fix bcs for BaseFormOperator on MixedFunctionSpace

### PR DESCRIPTION
**Pull request by pbrubeck**
# Description
This PR fixes several bugs related to applying `DirichletBC`s on `BaseFormOperators` involving `MixedFunctionSpace`.

1. `MixedInterpolator` was wrongly imposing bcs on `MixedFunctionSpaces` with repeated subspaces (which is a very common case).  For instance if `Z = V * V * Q`, the boundary condition `DirichletBC(Z.sub(0), "on_boundary")` would have been wrongly applied to both `Z.sub(0)` and `Z.sub(1)`, since they are equal to `V`. The fix is to check which bcs apply to each block of the matrix.

2. We also fix `assemble(interpolate(...), bcs=bcs)`, which was previously not passing the `bcs` on to the `Interpolator`.

3. Fixes https://github.com/firedrakeproject/firedrake/issues/4125, where we attempted to apply a DirichletBC to a bilinear form with different test and trial spaces. The fix is to zero the rows or columns corresponding to the BC nodes.

In addition, we also need to create jax/torch tensors with flattened `u.dat.data`. Suppose `u` comes from a `MixedFunctionSpace` then `torch.tensor(u.dat.data)` will receive a list of `numpy.arrays` and turn it into the columns of a matrix, which is an unexpected and undesired behaivor. A MixedFunctionSpace with subspaces of different dimension cannot be turned into a matrix in that way, as the columns would have different dimensions. Therefore `u.dat.data` must be flattened into a long vector.

### CODE REVIEW

File: tests/firedrake/external_operators/test_jax_operator.py [hunk 1]
```diff
@@ -245,3 +247,31 @@ def test_solve(mesh, V):
...
     assert err_point_expr < 1.0e-09
+
+
+def test_mixed_space_bcs():
+    mesh = UnitIntervalMesh(4)
+    V = FunctionSpace(mesh, "CG", 1)
+    W = V * V
+
+    test = TestFunction(W)
+    bcs = [DirichletBC(W.sub(0), Constant(1), 2),
```
Comment (pbrubeck) [inline tests/firedrake/external_operators/test_jax_operator.py]:
```suggestion
    bcs = [DirichletBC(W.sub(0), Constant(1), 1),
```

File: firedrake/assemble.py [hunk 2]
```diff
@@ -408,23 +408,32 @@ def assemble(self, tensor=None, current_state=None):
         in a post-order fashion and evaluating the nodes on the fly.
 
         """
+        def parent_space(V):
+            return V.parent or V
+
         def visitor(e, *operands):
             t = tensor if e is self._form else None
-            return self.base_form_assembly_visitor(e, t, *operands)
+            # Deal with 2-form bcs inside the visitor
```
Comment (connorjward) [inline firedrake/assemble.py:413]: Why do we deal with 2-form bcs here and 1-form bcs elsewhere? That seems like it will make the code very confusing to follow.

File: firedrake/assemble.py [hunk 3]
```diff
@@ -1446,7 +1453,11 @@ def _apply_bc(self, tensor, bc, u=None):
             V = V.parent
         index = 0 if V.index is None else V.index
         space = V if V.parent is None else V.parent
-        if isinstance(bc, DirichletBC):
+        if spaces[0] != spaces[1] and isinstance(bc, DirichletBC):
```
Comment (connorjward) [inline firedrake/assemble.py]: I think it would be clearer to have:
```py
if isinstance(bc, DirichletBC):
  if spaces[0] != spaces[1]:
    return  # with a useful comment

  ...
```

File: firedrake/interpolation.py [hunk 4]
```diff
@@ -1703,7 +1703,11 @@ def __init__(self, expr, V, bcs=None, **kwargs):
         super(MixedInterpolator, self).__init__(expr, V, bcs=bcs, **kwargs)
         expr = self.ufl_interpolate
         self.arguments = expr.arguments()
-        rank = len(self.arguments)
+        # Get the primal spaces
+        spaces = tuple(a.function_space().dual() if isinstance(a, Coargument) else a.function_space()
+                       for a in self.arguments)
+        # We need a stricter equality test for indexed MixedFunctionSpace
+        space_equals = lambda V1, V2: V1 == V2 and V1.parent == V2.parent and V1.index == V2.index
```
Comment (connorjward) [inline firedrake/interpolation.py:1711]: This should probably be implemented inside the appropriate `__eq__`.

Comment (pbrubeck) [reply firedrake/interpolation.py:1711]: That is an API breaking change, it should not go as part of this PR

File: firedrake/assemble.py (hunk 2 above)
Comment (pbrubeck) [reply firedrake/assemble.py:413]: The comment is to clarify that we were already treating bcs differently depending on the rank of the form. This separate treatment making the code much more easy to follow, as the 1-form case would require imposing bcs on every return statement in the visitor.

File: firedrake/interpolation.py (hunk 4 above)
Comment (connorjward) [reply firedrake/interpolation.py:1711]: But it's a bug, so arguably it's an API-fixing change

File: firedrake/assemble.py (hunk 2 above)
Comment (pbrubeck) [reply firedrake/assemble.py:413]: Matrices need the bcs at the moment of assembly, as we use masked LGMaps to not write on the boundary nodes. Functions can be easily modified post-assembly.

File: firedrake/interpolation.py (hunk 4 above)
Comment (pbrubeck) [reply firedrake/interpolation.py:1711]: Alright, I'll try this. If this breaks any tests, then I'll prefer to open a new PR

Comment (pbrubeck) [reply firedrake/interpolation.py:1711]: This seems quite involved. In `MixedFunctionSpace.__eq__` we do a subspace comparison. If we let `FunctionSpace.__eq__` also check for `self.parent == other.parent` on the parent `MixedFunctionSpace`, we create an infinte recursion.

File: firedrake/assemble.py (hunk 2 above)
Comment (connorjward) [reply firedrake/assemble.py:413]: It seems like a strange place to be filtering the BCs. Could we not do this when we make a `TwoFormAssembler` object?

File: firedrake/interpolation.py (hunk 4 above)
Comment (connorjward) [reply firedrake/interpolation.py:1711]: OK. If you create an issue for this and leave a TODO comment here pointing at it then I'm happy.

File: firedrake/interpolation.py [hunk 5]
```diff
@@ -1703,7 +1703,11 @@ def __init__(self, expr, V, bcs=None, **kwargs):
         super(MixedInterpolator, self).__init__(expr, V, bcs=bcs, **kwargs)
         expr = self.ufl_interpolate
         self.arguments = expr.arguments()
-        rank = len(self.arguments)
+        # Get the primal spaces
+        spaces = tuple(a.function_space().dual() if isinstance(a, Coargument) else a.function_space()
+                       for a in self.arguments)
+        # We need a stricter equality test for indexed MixedFunctionSpace
```
Comment (pbrubeck) [inline firedrake/interpolation.py]:
```suggestion
        # TODO consider a stricter equality test for indexed MixedFunctionSpace
        # See https://github.com/firedrakeproject/firedrake/issues/4668
```

File: firedrake/assemble.py (hunk 2 above)
Comment (pbrubeck) [reply firedrake/assemble.py:413]: I'll see if this works without filtering

File: firedrake/assemble.py [hunk 6]
```diff
@@ -408,23 +408,32 @@ def assemble(self, tensor=None, current_state=None):
...
+            return V.parent or V
+
         def visitor(e, *operands):
             t = tensor if e is self._form else None
-            return self.base_form_assembly_visitor(e, t, *operands)
+            # Deal with 2-form bcs inside the visitor
+            rank = len(self._form.arguments())
+            if (rank == 2 and isinstance(e, ufl.BaseForm) and len(e.arguments()) == rank):
+                args_fs = set(arg.function_space().dual() if isinstance(arg, ufl.Coargument) else arg.function_space() for arg in e.arguments())
+                bcs = tuple(bc for bc in self._bcs if parent_space(bc.function_space()) in args_fs)
```
Comment (pbrubeck) [inline firedrake/assemble.py]:
```suggestion
            if len(self._form.arguments()) == 2:
                bcs = self._bcs
```

File: firedrake/assemble.py [hunk 7]
```diff
@@ -408,23 +408,30 @@ def assemble(self, tensor=None, current_state=None):
...
+
         def visitor(e, *operands):
             t = tensor if e is self._form else None
-            return self.base_form_assembly_visitor(e, t, *operands)
+            # Deal with 2-form bcs inside the visitor
+            if len(self._form.arguments()) == 2:
+                bcs = self._bcs
+            else:
+                bcs = ()
+            return self.base_form_assembly_visitor(e, t, bcs, *operands)
```
Comment (pbrubeck) [inline firedrake/assemble.py]:
```suggestion
            return self.base_form_assembly_visitor(e, t, self._bcs, *operands)
```

File: firedrake/assemble.py (hunk 2 above)
Comment (pbrubeck) [reply firedrake/assemble.py:413]: we don't need the filtering, seems like that's done in `TwoFormAssembler`

File: firedrake/assemble.py [hunk 8]
```diff
@@ -408,23 +408,27 @@ def assemble(self, tensor=None, current_state=None):
...
 
         """
+        def parent_space(V):
+            return V.parent or V
+
         def visitor(e, *operands):
             t = tensor if e is self._form else None
-            return self.base_form_assembly_visitor(e, t, *operands)
+            # Deal with 2-form bcs inside the visitor
+            bcs = self._bcs if isinstance(e, ufl.BaseForm) and len(e.arguments()) == 2 else ()
```
Comment (connorjward) [inline firedrake/assemble.py:414]: Why is this necessary?

**Review by connorjward — COMMENTED** [review]
Other than this query I think I'm happy.

File: firedrake/assemble.py (hunk 8 above)
Comment (pbrubeck) [reply firedrake/assemble.py:414]: We were always ignoring bcs for 1-form assembly already, this is just making the comments above more explicit

File: firedrake/assemble.py [hunk 9]
```diff
@@ -408,23 +408,27 @@ def assemble(self, tensor=None, current_state=None):
         in a post-order fashion and evaluating the nodes on the fly.
 
         """
+        def parent_space(V):
+            return V.parent or V
```
Comment (connorjward) [inline firedrake/assemble.py]: This isn't used any more.

File: firedrake/assemble.py (hunk 8 above)
Comment (connorjward) [reply firedrake/assemble.py:414]: Yeah I guess this is better.

File: firedrake/assemble.py [hunk 10]
```diff
@@ -408,23 +408,27 @@ def assemble(self, tensor=None, current_state=None):
...
+
```
Comment (pbrubeck) [inline firedrake/assemble.py]:
```suggestion
```
